test_util/aws.py

```diff
@@ -179,77 +243,59 @@ def create(cls, stack_name, template_url, public_agents, private_agents,
         self.stack = self.boto_wrapper.resource('cloudformation').Stack(self.stack.stack_id)
+        # check for things that have to delete explicitly
+        log.info('Starting deletion of CF stack')
+        self.delete_exhibitor_s3_bucket()
         self.stack.delete()
-        self.empty_and_delete_s3_bucket_from_stack()
-
-    def empty_and_delete_s3_bucket_from_stack(self):
-        bucket_id = self.stack.Resource('ExhibitorS3Bucket').physical_resource_id
-        s3 = self.boto_wrapper.resource('s3')
-        bucket = s3.Bucket(bucket_id)
-        log.info('Starting bucket {} deletion'.format(bucket))
-        all_objects = bucket.objects.all()
-        obj_count = len(list(all_objects))
-        if obj_count > 0:
-            assert obj_count == 1, 'Expected one object in Exhibitor S3 bucket but found: ' + str(obj_count)
-            exhibitor_object = list(all_objects)[0]
-            log.info('Trying to delete object from bucket: {}'.format(repr(exhibitor_object)))
-            exhibitor_object.delete()
-        log.info('Trying deleting bucket {} itself'.format(bucket))
-        bucket.delete()
-        log.info('Delete successfully triggered for {}'.format(self.stack.stack_name))
 
     def get_master_ips(self, refresh=False):
-        return self.get_hosts_cached('MasterServerGroup', refresh=refresh)
+        masters_fn = functools.partial(
+            self.boto_wrapper.get_auto_scaling_hosts,
+            self.stack.Resource('MasterServerGroup').physical_resource_id)
+        return self._cached('masters', masters_fn, refresh=refresh)
 
     def get_public_agent_ips(self, refresh=False):
-        return self.get_hosts_cached('PublicSlaveServerGroup', refresh=refresh)
+        public_agents_fn = functools.partial(
+            self.boto_wrapper.get_auto_scaling_hosts,
+            self.stack.Resource('PublicSlaveServerGroup').physical_resource_id)
+        return self._cached('public_agents', public_agents_fn, refresh=refresh)
 
     def get_private_agent_ips(self, refresh=False):
-        return self.get_hosts_cached('SlaveServerGroup', refresh=refresh)
+        private_agents_fn = functools.partial(
+            self.boto_wrapper.get_auto_scaling_hosts,
+            self.stack.Resource('SlaveServerGroup').physical_resource_id)
+        return self._cached('private_agents', private_agents_fn, refresh=refresh)
+
+    def delete_exhibitor_s3_bucket(self):
+        """ A non-empty S3 bucket cannot be deleted, so check to
+        see if it should be emptied first. If its non-empty, but
+        has more than one item, error out as the bucket is perhaps
+        not an exhibitor bucket and the user should be alerted
+        """
+        bucket = self.boto_wrapper.resource('s3').Bucket(
+            self.stack.Resource('ExhibitorS3Bucket').physical_resource_id)
+        log.info('Starting bucket {} deletion'.format(bucket))
+        all_objects = list(bucket.objects.all())
```

Comment:
Good to know, but in this specific case we actually want to get the resources in the list to potentially delete, so we do want a list